[stdlib] 0-ary tuples should be equatable

anayini opened the pull request:
Add an == method for comparing two empty tuples.

Resolves: SR-4172 (https://bugs.swift.org/browse/SR-4172)


<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Commit:
[stdlib] 0-ary tuples should be equatable

Add an == method for comparing two empty tuples.

Resolves: SR-4172 (https://bugs.swift.org/browse/SR-4172)

anayini commented:
@swift-ci Please smoke test

CodaFi reviewed (changes requested):
May as well do the full monty and edit the gyb template to make tuples quasi-`Comparable` too.  This is also missing `!=`.

Commit:
[stdlib] Add other comparison operators to 0-ary tuples

Add `!=`, `<`, `<=`, `>`, `>=` to 0-ary tuples

Resolves: SR-4172 (https://bugs.swift.org/browse/SR-4172)

anayini commented:
@CodaFi - Thanks.  Just to clarify edit the .gyb to add `!=`, `<`, `<=`, `>`, `>=`?

CodaFi commented:
You should edit the GYB template to account for the non-generic 0-ary tuple.  It should be possible to do this "in GYB"'s language so we don't litter this file with just this specific conformance.

anayini commented:
Gotcha - was considering that adding extra logic for the 0 case in the existing templated methods might it actually more complicated than having a specific extra section for 0 outside of that, but can do that if preferred

practicalswift commented:
👍 for the initiative/idea!

moiseev commented:
Thanks @anayini! I agree that trying to incorporate the `()` into an existing gyb is not worth the effort, especially given that the one-tuple case will still be missing.
Please add the missing `!=` and `Comparable` operators with doc comments.

moiseev commented on stdlib/public/core/Tuple.swift.gyb:
```diff
@@ -23,6 +23,14 @@ comparableOperators = [
 
 }%
 
+/// Two tuples each of arity 0 are equal
```
I would copy the heading from the 'generic' `==` on tuples, and later add a sentence that all empty-tuples are equal to each other as a note.

```swift
// Returns a Boolean value indicating whether the corresponding components of
/// two tuples are equal.
///
/// All zero-arity tuples are equal.
```

Similarly for the `<` etc...

anayini replied:
👍

anayini commented:
@moiseev - Updated to address your comments.

moiseev commented on stdlib/public/core/Tuple.swift.gyb:
```diff
@@ -23,6 +23,82 @@ comparableOperators = [
+/// Returns a Boolean value indicating whether any corresponding components of
+/// the two tuples are not equal.
+///
+/// Two tuples each of arity zero are never unequal
```
'never unequal' is a bit hard to understand (and there is a period missing ;-)). I'd go with the same note as for the `==`.

anayini replied:
Makes sense

moiseev commented:
I don't mind the non-gybbed version. @CodaFi , @airspeedswift what do you think?

Commit:
[stdlib] Address PR style comments

CodaFi commented:
I don't have any qualms with not using GYB here.  My thinking is that the file would be more resilient to future interface changes/tuple changes and we could update everything all at once.  But that's a lot of work that can be done later.  For now, LGTM.

anayini commented:
There a bit of more discussion here: https://bugs.swift.org/browse/SR-4172, regarding the size of this code and whether the `Comparable` operators are needed.  It also includes a helpful example for why the `==` operator could be useful to someone.

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

anayini commented:
Welp, I suck.  Looks like a failure in `IDE/complete_operators.swift`, which I guess I didn't run locally.  Will investigate at some point today

moiseev commented:
@anayini that's why we have CI ;-) Few people have enough resources and patience to run all the tests locally.

Commit:
[stdlib] Update IDE tests

anayini commented:
Long time since last update, sorry about that. - @moiseev I'm actually still pretty confused about how the IDE code completion tests run using FileCheck.  I've got something that I believe passes, but I'm definitely still confused about the semantics of the file.

anayini commented on test/IDE/complete_operators.swift:
```diff
@@ -40,7 +40,7 @@
 // RUN: %target-swift-ide-test -code-completion -source-filename=%s -code-completion-token=INFIX_15 | %FileCheck %s -check-prefix=NO_OPERATORS
 // RUN: %target-swift-ide-test -code-completion -source-filename=%s -code-completion-token=INFIX_16 | %FileCheck %s -check-prefix=NO_OPERATORS
-// RUN: %target-swift-ide-test -code-completion -source-filename=%s -code-completion-token=INFIX_17 | %FileCheck %s -check-prefix=NO_OPERATORS
+// RUN: %target-swift-ide-test -code-completion -source-filename=%s -code-completion-token=INFIX_17 | %FileCheck %s -check-prefix=VOID_OPERATORS
```
Should maybe rename `VOID_OPERATORS` to `VOID_INFIX_OPERATORS`?

anayini commented on test/IDE/complete_operators.swift:
```diff
@@ -259,9 +260,20 @@ func testInfix15<T: P where T.T == S2>() {
   x#^INFIX_17^#
 }
+
+// VOID_OPERATORS: Begin completions
```
Generally want some feedback as to where the `CHECK:` lines need to go.  Even before this PR seems like the `NO_OPERATORS-NOT` was in a strange place, but could be my misunderstanding.

moiseev replied:
FileCheck is part of LLVM. See [here](http://llvm.org/docs/CommandGuide/FileCheck.html) for the documentation.

moiseev replied:
Oh, I see how this is related to the change. `()` is same as `Void`, and therefore we get all the equality/comparison ops on the value of type `Void`. This might have other consequences...

anayini replied:
Yup.  Took me a bit to figure out, but yeah SourceKit code completion was offering suggestions of all the equality/comparison ops that were added, but it wasn't expecting any suggestions so this test was failing.

moiseev replied:
/cc @airspeedswift

moiseev commented:
@swift-ci Please smoke test

moiseev commented on test/IDE/complete_operators.swift:
```diff
@@ -244,6 +244,7 @@ func testInfix11() {
   S2#^INFIX_11^#
 }
 // NO_OPERATORS-NOT: Decl[InfixOperatorFunction]
+
```
Why the newline?

anayini replied:
Ah sorry - I can remove that.

anayini commented:
bump @airspeedswift 

I can fix the newline nits, but wanted to see if you had any other feedback if you have some time
:)

anayini commented:
@moiseev - anyone else we can get feedback from :)

I'll push up the whitespace fix, but was going to wait for other feedback in case its needed.

moiseev commented:
@anayini Sorry for the silence. Last I remember there were some discussions about it. We'll get back to you once things clear up. Thanks for your patience!

anayini commented:
@moiseev - Just checking in again :)

anayini commented:
@moiseev - Did ya'll come to a conclusion?  Should I close for now?

anayini committed:
Merge branch 'master' into bugfix/SR-4172

moiseev commented:
@swift-ci please smoke test

moiseev commented:
@swift-ci Please test source compatibility

moiseev commented:
@swift-ci Please Test Source Compatibility

anayini commented:
@moiseev - Can dive into the Source Compatibility Suite error log - do these tests ever flake?  If so, can you rerun?

jrose-apple commented:
This one's a known issue that we just XFAILed. I'll kick off another run for you.

@swift-ci Please test source compatibility

anayini commented:
Hm seems like this time its an `IndexError: list index out of range` from 
 `xfail['branch'][swift_branch].split()[0]`

jrose-apple commented:
Finally back to normal!

@swift-ci Please test source compatibility

anayini commented:
😄

anayini commented:
@moiseev - Are we good to merge?

moiseev commented:
@anayini, do you mind sending an email to swift-evolution with a proposal to refine https://github.com/apple/swift-evolution/blob/master/proposals/0015-tuple-comparison-operators.md to cover the empty tuple case?

anayini commented:
@moiseev - Sorry for the delay - was getting ready for a long OOTO and am now out of the country until the end of the month.  Will be able to check in on this periodically.

anayini commented:
@moiseev - Looks like it was accepted to be posted on swift-evolution.  What are the next steps?

moiseev commented:
@anayini Thanks for the contribution and your patience!
